### PR TITLE
TIP-709: indexing format to be used in Elasticsearch

### PR DESCRIPTION
**Pull request by jjanvier**
## Overall plan

Here is the second PR of a series which goal is to index our products in ElasticSearch:

1. introduce a simple and stupid ElasticSearch client provided as a Symfony service ([here](https://github.com/jjanvier/pim-community-dev/pull/52))
2. define an indexing format for our products (new product normalizers coming...) (this PR)
3. index our products when they are saved
4. turn the first integration tests of the PQB green

## This PR

- introduces the indexing format that will be used in Elasticsearch (based on the standard format)
- moves the ES filters documentation to the proper folder
- attempts to fix the integration tests related to indexing/storage and standard normalization

At some point, it would be interesting to clean our normalizer/serializer stack to avoid to have to add new "empty" normalizers (some of them just composer with the standard, nothing more). This is planned in another story \o/.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | Y
| Changelog updated                 | -
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | 
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * Also, prices are indexed by currency. See the method "normalizePriceCollectionData" below.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductValueNormalizer implements NormalizerInterface
+{
+    /** @var NormalizerInterface */
+    protected $stdNormalizer;
```
Comment (anaelChardan) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:48]: Just a question, why not standardNormalizer ? In my head, below when I see stdNormalizer and after stdValues it rings like stdClass values :/.

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     *  "EUR" => array:2 [
+     *      "amount" => "-56.53"
+     *      "currency" => "EUR"
+     *  ]
+     *
+     * @param array $stdPrices
+     *
+     * @return array
+     */
+    private function normalizePriceCollectionData(array $stdPrices)
```
Comment (anaelChardan) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:118]: same

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php (hunk 1 above)
Comment (anaelChardan) [reply src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:48]: Or it is and I would prefer stdClassNormalizer

File: src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/ProductValuesNormalizerSpec.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(ProductValuesNormalizer::class);
+    }
+
+    function it_is_a_normalizer()
+    {
+        $this->shouldImplement('Symfony\Component\Serializer\Normalizer\NormalizerInterface');
```
Comment (anaelChardan) [inline src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/ProductValuesNormalizerSpec.php:31]: When do we use MyInterface::class and when do we use FQCN ?

File: src/Pim/Component/Catalog/tests/integration/Normalizer/NormalizedProductCleaner.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        return $sortedValues;
+    }
+
+    /**
+     * @param mixed $array
+     * @param int   $sort_flags
+     *
+     * @return bool
+     */
+    private function ksortRecursive(&$array, $sort_flags = SORT_REGULAR)
```
Comment (anaelChardan) [inline src/Pim/Component/Catalog/tests/integration/Normalizer/NormalizedProductCleaner.php:115]: Should be a static one :)

File: src/Pim/Component/Catalog/Normalizer/Indexing/DateTimeNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Normalize a DateTime as ISO-8601 (with the timezone) to the indexing format.
+ * See https://en.wikipedia.org/wiki/ISO_8601
+ *
+ * Example: 2017-06-13T12:07:58+00:00
+ *
+ * This format is based on the standard format.
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
```
Comment (BitOne) [inline src/Pim/Component/Catalog/Normalizer/Indexing/DateTimeNormalizer.php:15]: Not sure that this address exists ;p

**Review by anaelChardan — COMMENTED** [review]
Only questions

File: src/Pim/Bundle/CatalogBundle/Resources/config/serializers_indexing.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    pim_catalog.normalizer.indexing.file.class: Pim\Component\Catalog\Normalizer\Indexing\FileNormalizer
+    pim_catalog.normalizer.indexing.attribute_option.class: Pim\Component\Catalog\Normalizer\Indexing\AttributeOptionNormalizer
+services:
+    pim_catalog.normalizer.indexing.product:
+        class: '%pim_catalog.normalizer.indexing.product.class%'
+        arguments:
+            - '@pim_catalog.normalizer.indexing.product.properties'
+            - '@pim_catalog.normalizer.indexing.product.associations'
+        tags:
+            - { name: pim_serializer.normalizer, priority: 90 }
```
Comment (BitOne) [inline src/Pim/Bundle/CatalogBundle/Resources/config/serializers_indexing.yml:19]: Shouldn't we have a dedicated normalizer instead on relying on the generic (and over-used) pim_serializer ?

File: src/Pim/Component/Catalog/Normalizer/Indexing/FileNormalizer.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Pim\Component\Catalog\Normalizer\Indexing;
+
+use Akeneo\Component\FileStorage\Model\FileInfoInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalize files to the indexing format.
+ * This format is based on the standard format.
```
Comment (BitOne) [inline src/Pim/Component/Catalog/Normalizer/Indexing/FileNormalizer.php:10]: Why?

File: src/Pim/Component/Catalog/Normalizer/Indexing/AttributeOptionNormalizer.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        return $this->stdNormalizer->normalize($attributeOption, $format, $context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return $data instanceof AttributeOptionInterface && 'indexing' === $format;
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Normalizer/Indexing/AttributeOptionNormalizer.php]: maybe change 'indexing' for something that is not a verb like 'indexed'/'indexation' ? Not mandatory just a proposition

**Comment by BitOne** [comment]
As the indexing format main purpose is to serve search queries, I don't get why it's based on the standard format which has not the same purpose.

More precisely, we shouldn't even have normalizer for field for which we don't have a filter. For example, the file normalizer.

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php [hunk 9]
```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Pim\Component\Catalog\Normalizer\Indexing\Product;
+
+use Pim\Component\Catalog\AttributeTypes;
+use Pim\Component\Catalog\Model\ProductValueInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalize a product value to the indexing format.
```
Comment (SamirBoulil) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:10]: 👍

**Review by BitOne — CHANGES_REQUESTED** [review]

File: src/Pim/Component/Catalog/tests/integration/Normalizer/Indexing/ProductIndexingIntegration.php [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+use Akeneo\Test\Integration\Configuration;
+use Akeneo\Test\Integration\TestCase;
+use Pim\Component\Catalog\tests\integration\Normalizer\NormalizedProductCleaner;
+
+/**
+ * Integration tests to verify data from database are well formatted in the indexing format
+ */
+class ProductIndexingIntegration extends TestCase
+{
+    protected function getConfiguration()
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/tests/integration/Normalizer/Indexing/ProductIndexingIntegration.php:14]: `{@inheritdoc}`

**Review by damien-carcel — APPROVED** [review]

**Comment by jjanvier** [comment]
@BitOne 

> As the indexing format main purpose is to serve search queries, I don't get why it's based on the standard format which has not the same purpose.

2 reasons: 
- because even if they are not indexed the same way, the data really are the same between our different formats (standard, indexing, storage). So I think it could be relevant to be based on the standard format, at least for the data.
- for commodity because it helped us to to code faster (not a good reason, we agree, please see below)

> More precisely, we shouldn't even have normalizer for field for which we don't have a filter. For example, the file normalizer.

I completely agree. We have all these useless normalizers currently because we are doing some really really unusual and weird things with the Symfony serializers/normalizers. Please see [this conversation](https://github.com/akeneo/pim-community-dev/pull/5418#discussion_r95167513) for instance. (It would easier to explain IRL)

My plan was to:
- first, introduce quickly this indexing format (the most important class in the PR is the integration test imo). Using the standard normalizers is just a way to achieve that, not a goal or a will.
- then, in another PR, review all our serialzers/normalizers (versioning, standard, ui, indexing, storage) stack to not be forced to write dirty things like what I had to do here. Which would allow to remove all these useless normalizers.

**Comment by BitOne** [comment]
@jjanvier:

So if I understand correctly, this is not the final indexing format, but just a quick way to have some data in Elasticsearch.

If that's the case, your plan is missing a "Build the real final indexing format" as step 2.

And yes, the normalizer stuff is a issue, but a different one.

**Comment by jjanvier** [comment]
@BitOne it should be the final indexing format (unless we missed some stuff)
but maybe not the way to build it

IRL this morning?

**Comment by jjanvier** [comment]
Seen IRL with Benoit.
We'll do another way.
Each filter will normalizes its small part of the product.

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php (hunk 1 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:48]: ok $standardNormalizer would be better
just a copy/paste I guess

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php (hunk 2 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php]: same what?

Comment (jjanvier) [reply src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:118]: $standardPrices would be better

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php (hunk 1 above)
Comment (anaelChardan) [reply src/Pim/Component/Catalog/Normalizer/Indexing/Product/ProductValueNormalizer.php:48]: Thanks seigneur jj

File: src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/ProductValuesNormalizerSpec.php (hunk 3 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/spec/Normalizer/Indexing/Product/ProductValuesNormalizerSpec.php:31]: use MyInterface::class whenever you can
it's a copy/paste I guess

File: src/Pim/Component/Catalog/tests/integration/Normalizer/NormalizedProductCleaner.php (hunk 4 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/tests/integration/Normalizer/NormalizedProductCleaner.php:115]: agree

File: src/Pim/Bundle/CatalogBundle/Resources/config/serializers_indexing.yml (hunk 6 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Resources/config/serializers_indexing.yml:19]: yep agree

**Pull request closed without merging**